cmd/describe/machinepool/cmd.go

```diff
@@ -29,61 +29,68 @@ import (
-	Aliases: []string{"machine-pool"},
-	Short:   "Show details of a machine pool on a cluster",
-	Long:    "Show details of a machine pool on a cluster.",
-	Example: `  # Show details of a machine pool named "mymachinepool"" on a cluster named "mycluster"
-  rosa describe machinepool --cluster=mycluster --machinepool=mymachinepool`,
-	Run:  run,
-	Args: cobra.MaximumNArgs(1),
-}
+const (
+	use     = "machinepool"
+	alias   = "machine-pool"
+	short   = "Show details of a machine pool on a cluster"
+	long    = "Show details of a machine pool on a cluster."
+	example = `  # Show details of a machine pool named "mymachinepool"" on a cluster named "mycluster"
+  rosa describe machinepool --cluster=mycluster --machinepool=mymachinepool`
+)
 
-var args struct {
-	machinePool string
-}
+func NewDescribeMachinePoolCommand() *cobra.Command {
+	options := NewDescribeMachinepoolUserOptions()
+	cmd := &cobra.Command{
+		Use:     use,
+		Short:   short,
+		Long:    long,
+		Aliases: []string{alias},
+		Example: example,
+		Args:    cobra.NoArgs,
+		Run:     rosa.DefaultRunner(rosa.RuntimeWithOCM(), DescribeMachinePoolRunner(options)),
+	}
 
-func init() {
-	flags := Cmd.Flags()
-	flags.SortFlags = false
-	ocm.AddClusterFlag(Cmd)
-	output.AddFlag(Cmd)
+	flags := cmd.Flags()
 	flags.StringVar(
-		&args.machinePool,
+		&options.machinepool,
 		"machinepool",
 		"",
 		"Machine pool of the cluster to target",
 	)
-}
 
-func run(cmd *cobra.Command, argv []string) {
-	r := rosa.NewRuntime().WithAWS().WithOCM()
-	defer r.Cleanup()
-	err := runWithRuntime(r, cmd, argv)
-	if err != nil {
-		r.Reporter.Errorf(err.Error())
-		os.Exit(1)
-	}
+	output.AddFlag(cmd)
+	ocm.AddClusterFlag(cmd)
+	return cmd
 }
 
-func runWithRuntime(r *rosa.Runtime, cmd *cobra.Command, argv []string) error {
-	// Allow the use also directly the machine pool id as positional parameter
-	if len(argv) == 1 && !cmd.Flag("machinepool").Changed {
-		args.machinePool = argv[0]
-	}
-	if args.machinePool == "" {
-		return fmt.Errorf("You need to specify a machine pool name")
-	}
-	clusterKey := r.GetClusterKey()
-	cluster := r.FetchCluster()
-	if cluster.State() != cmv1.ClusterStateReady {
-		r.Reporter.Errorf("Cluster '%s' is not yet ready", clusterKey)
-		os.Exit(1)
-	}
-	isHypershift := cluster.Hypershift().Enabled()
+func DescribeMachinePoolRunner(userOptions DescribeMachinepoolUserOptions) rosa.CommandRunner {
+	return func(_ context.Context, runtime *rosa.Runtime, cmd *cobra.Command, argv []string) error {
+		options, err := NewDescribeMachinepoolOptions()
+		if err != nil {
+			return fmt.Errorf("unable to create options for Describe Machinepool runner: %s", err)
+		}
+		// Allow the use also directly the machine pool id as positional parameter
+		if len(argv) == 1 && !cmd.Flag("machinepool").Changed {
+			userOptions.machinepool = argv[0]
+		} else {
+			err := cmd.ParseFlags(argv)
+			if err != nil {
+				return fmt.Errorf("unable to parse flags: %s", err)
+			}
+		}
+		options.Bind(userOptions)
```

Comment:
@hunterkepley I think it makes sense for the `Bind` method to return an `error` if the user hasn't supplied enough fields or fields that contradict each other. Essentially `Bind` validates the options the user has supplied.

So for all validation of user options it would look like:

```
err  := options.Bind(userOptions)
if err != nil {
 return err
}
```

We can then extend the validation on the options without having to touch the runner.  We can also unit test fairly easily option validation scenarios.

Reply:
Sounds good

Reply:
Updated tests and added validation + error return to bind